dask/utils.py

```diff
@@ -1229,6 +1229,18 @@ def get_scheduler_lock(collection=None, scheduler=None):
 
     if actual_get == multiprocessing.get:
         return multiprocessing.get_context().Manager().Lock()
+    else:
+        # if this is a distributed client with processes, we also need to lock
+        try:
+            from distributed.utils import get_mp_context
+            from distributed.worker import get_client
+
+            client = get_client()
+        except (ImportError, ValueError):
+            pass
+        else:
+            if actual_get == client.get and getattr(client.cluster, "processes", None):
```

Comment:
this only works when using a cluster directly, if I run with:

```python
from __future__ import annotations

import os
import contextlib
import pathlib
import sys
import tempfile

import pandas as pd

from distributed import Client, LocalCluster, Lock

import dask.dataframe as dd


@contextlib.contextmanager
def _tmp_path():
    with tempfile.TemporaryDirectory() as tmp_dir:
        yield pathlib.Path(tmp_dir)


def main():
    with (
        Client(scheduler_file=os.environ["SCHEDULER_FILE"]) as client,
        _tmp_path() as tmp_path,
    ):
        file = tmp_path / "input.h5"
        dd.from_pandas(pd.DataFrame(range(100)), 100).to_hdf(file, key="/demo*")


if __name__ == "__main__":
    sys.exit(main())
```

then I run a cluster manually with:
```sh
dask scheduler --scheduler-file=scheduler.json
```
```
dask worker --scheduler-file=scheduler.json
```
```
dask worker --scheduler-file=scheduler.json
```
```
dask worker --scheduler-file=scheduler.json
```
```sh
SCHEDULER_FILE=scheduler.json python demo.py
```

I still get:

```
    unable to lock file, errno = 11, error message = 'Resource temporarily unavailable'

End of HDF5 error back trace

Unable to open/create file '/tmp/tmp1sd69d9i/input.h5'
```

Reply:
This is now resolved.